libraries/SITL/SIM_JEDEC.cpp

```diff
@@ -0,0 +1,177 @@
+#include "SIM_JEDEC.h"
+
+#include <errno.h>
+#include <unistd.h>
+
+#include <AP_HAL_SITL/AP_HAL_SITL.h>
+
+using namespace SITL;
+
+extern HAL_SITL& hal;
+
+void JEDEC::open_storage_fd()
+{
+    if (storage_fd != -1) {
+        AP_HAL::panic("Should not have been called");
+    }
+    
+    const char *filepath = filename();
+    uint32_t flags = O_RDWR|O_CLOEXEC;
+    storage_fd = open(filepath, flags, 0777);
+
+    if (storage_fd == -1) {
+        flags |= O_CREAT;
+        storage_fd = open(filepath, flags, 0777);
+    }
+
+    // still -1 ??
+    if (storage_fd == -1) {
+        AP_HAL::panic("open(%s): %s", filepath, strerror(errno));
+    }
+    if (ftruncate(storage_fd, storage_size()) != 0) {
+        AP_HAL::panic("truncate(%s): %s", filepath, strerror(errno));
+    }
+}
+
+int JEDEC::rdwr(uint8_t count, SPI::spi_ioc_transfer *&tfrs)
+{
+    if (storage_fd == -1) {
+        open_storage_fd();
+    }
+
+    // commands:
+    static const uint8_t JEDEC_RDID             = 0x9f;
+    static const uint8_t JEDEC_READ             = 0x03;
+    static const uint8_t JEDEC_WREN             = 0x06;
+    static const uint8_t JEDEC_WRITE            = 0x02;
+    static const uint8_t JEDEC_RDSR             = 0x05;
+    static const uint8_t JEDEC_SECTOR4_ERASE    = 0x20;
+    static const uint8_t JEDEC_BULK_ERASE       = 0xC7;
+    static const uint8_t JEDEC_BLOCK64_ERASE    = 0xD8;
+
+    for (uint8_t i=0; i<count; i++) {
+        SPI::spi_ioc_transfer &tfr = tfrs[i];
+        uint8_t *tx_buf = (uint8_t*)(tfr.tx_buf);
+        uint8_t *rx_buf = (uint8_t*)(tfr.rx_buf);
+
+        switch (state) {
+        case State::WAITING: {
+            // find a command
+            uint8_t command = tx_buf[0];
+            switch (command) {
+            case JEDEC_RDID:
+                state = State::READING_RDID;
+                break;
+            case JEDEC_READ:
+                xfr_addr = tx_buf[1] << 16 | tx_buf[2] << 8 | tx_buf[3];
```

Comment:
We should probably add some sanity checks in here that we've actually got that many bytes in the transfer.

Similarly elsewhere

Reply:
Done! Thanks @hendjoshsr71 for helping me do this.